### PR TITLE
Feature: include the pharmacy's phone number in the Rx chart note

### PR DESCRIPTION
**Pull request by LiamStanziani**
# Summary

When a prescription is faxed, the note posted to the patient's chart now includes the pharmacy's
phone number alongside the fax number:

```
[Rx faxed to Example Pharmacy Fax#: 5555550100 Tel: 5555550199 prescribed by Testerson, Testy, Jan 01, 2026, 09:00 AM]
```

The `Tel:` label and the phone1/phone2 composition deliberately match the pharmacy address block
already printed on the prescription, so the note shows the same numbers the pharmacy was sent.

Full write-up, screenshots and test evidence: **[Pharmacy Phone Number in Rx Chart Note Feature](https://docs.google.com/document/d/1cu7i9flHbsyYsXdLhYhJ-Hdbo8Ek0cx0o6SXn31fEBk/edit?usp=sharing)**

Fixes: [Include pharmacy's phone # in rx chart note](https://trello.com/c/STmlofNK/50-include-pharmacys-phone-in-rx-chart-note)

## Problem

Requested by a Magenta MD:

> May I possibly suggest a tinkering to the way prescriptions are posted to the chart? Currently
> it'll show the fax number, but is it possible to also post the phone number? […] Sometimes I find
> I need to call the pharmacy, to clarify things, and it would be so much easier if the phone number
> were just there when the faxed prescription is posted.

The chart note recorded the fax number only. A prescriber needing to call the pharmacy — to clarify a
dose, a substitution, a rejected claim — had to leave the chart, open the pharmacy picker and look the
number up. The one number on screen was the one number they couldn't dial.

## Solution

- **`RxPharmacyData.composePharmacyPhone(PharmacyInfo)`** — a single shared helper that joins `phone1`
  and `phone2` with one space, skipping whichever is absent. It lives on `RxPharmacyData` because that
  is the pharmacy class both call sites already import and instantiate.
- **`RxPrintPreview2Action`** uses it for the printed address block and exposes the result as the
  `pharmacyPhone` request attribute.
- **`PrintPreview.js` / `PrintPreview.jsp`** thread that attribute through and emit the `Tel:` segment.
- **`ViewScript2.jsp`** — the legacy Rx preview holds a second copy of this note string. It is
  unreachable today, but leaving one of two copies behind is how formats silently diverge.

**Behaviour**

| Case | Note segment |
| :---- | :---- |
| `phone1` + `phone2` | `Fax#: … Tel: (416) 269-4820 4165550000` |
| `phone1` only | `Fax#: … Tel: (416) 269-4820` |
| `phone2` only | `Fax#: … Tel: 4165550000` |
| Neither | `Fax#: …` — no `Tel:` segment, exactly as today |

Numbers are rendered **verbatim** — leading/trailing whitespace is trimmed and embedded newlines
collapse to a space, but nothing is reformatted. Stored pharmacy phone numbers are unvalidated free
text in mixed

The change is **self-suppressing**: a pharmacy with no phone on file gets no `Tel:` segment at all, so
there is no property flag. (The separate space-restoring commit does change that line by one
character — see the notes for reviewers below.)

**UI Before:**
image

**UI After:**
image

## Validation

- **Manual UI testing** across five pharmacies covering every branch: `phone1` only, both numbers in
  mixed formats, `phone2` only, no phones with an apostrophe in the name, and embedded double quotes
  in the phone field. Screenshots in the linked doc.
- **8 unit tests** — `RxPharmacyDataPhoneTest` covers all four composition cases plus null pharmacy,
  whitespace-only input, newline collapsing and verbatim passthrough. Verified they actually execute
  under the modern surefire include patterns, not just that the build is green.
- **Both changed JSPs precompiled with Jasper** to `.class`. This matters for `ViewScript2.jsp`: it is
  dead code, so a scriptlet error there would never surface at runtime.
- **Backport rehearsed** — all five commits cherry-picked onto `release/2026-03-17` in a scratch
  worktree. Four apply clean; the legacy-JSP commit takes a one-line conflict resolution, documented
  in the linked doc.

## Additional notes for reviewers

- **Fixes a latent bug on the printed Rx.** `setupPharmacyAddress` had an unguarded `getPhone1()`, so a
  pharmacy with no `phone1` printed `Tel: null`, and one with only `phone2` printed
  `Tel: null 4165550000`. Routing both through the shared helper removes those. This changes the
  *prescription*, not just the note — called out deliberately rather than slipped in.
- **One commit is not part of the feature.** The last commit restores a missing space before the
  timestamp (`… prescribed by Name,Jan 01 …` → `… prescribed by Name, Jan 01 …`). `ViewScript2.jsp`,
  which this path was ported from, has the space — it was dropped in the port. Isolated so it can be
  reverted independently if wanted.
- **The timestamp *format* is deliberately untouched.** `PrintPreview.js` builds its own timestamp via
  `toLocaleString` and ignores the server-computed `timeStamp` attribute, which has zero consumers.
  Do not "fix" this by switching to the server value:
  [MagentaHealth#152](https://github.com/MagentaHealth/Open-O/pull/152) removed `timeZone: 'UTC'` from
  exactly these lines so the note renders in local time, and moving to `SimpleDateFormat` would hand
  the timestamp back to the server's timezone. The real follow-up is to delete the dead attribute.
- **Quote handling.** Quotes in pharmacy fields have broken this note before (#2483, and
  MagentaHealth#144). `pharmacyPhone` is a new value flowing into the same JS string literal, so it is
  encoded at the output point (`${e:forJavaScript(...)}` / `Encode.forJavaScript`) and explicitly
  tested with embedded quotes.

## Security

- OWASP encoding at every JS output point; the value is stored raw and encoded per context, matching
  the existing `pharmacyName` / `pharmacyFax` convention.
- No SQL — reuses `RxPharmacyData.getPharmacy()`.
- Privilege checks unchanged; `RxWriteToEncounter2Action` still enforces `_rx` write.
- No new logging.

---

## Summary by Sourcery

Include composed pharmacy phone numbers in faxed prescription chart notes and printed Rx address blocks while correcting chart note timestamp spacing.

New Features:
- Add pharmacy phone number to the "Rx faxed to" chart note when available.

Bug Fixes:
- Prevent printed prescriptions from showing "Tel: null" or stray separators when pharmacy phone fields are missing.

Enhancements:
- Introduce a shared helper to compose pharmacy phone1/phone2 into a single display value used across Rx views.
- Extend both modern and legacy Rx print preview flows to pass and safely encode pharmacy phone data in JavaScript-driven notes.

Tests:
- Add unit test coverage for pharmacy phone composition across null, empty, whitespace, newline, and mixed-format cases.

---
## Summary by cubic
Include the pharmacy phone number in the “Rx faxed to …” chart note so providers can call without leaving the chart. Previously the note showed only the fax; now it appends “Tel: <phone1> <phone2>” verbatim and skips the segment when no phone exists; printed Rxs no longer show “Tel: null,” and the note restores a missing space before the timestamp.

- Adds `RxPharmacyData.composePharmacyPhone(PharmacyInfo)` to join `phone1`/`phone2`, trim, and collapse newlines; returns "" when both are blank; documents with `@since`.
- Uses the helper in `RxPrintPreview2Action` to render the printed address block and exposes `pharmacyPhone` to `PrintPreview.jsp`/`PrintPreview.js`; updates `ViewScript2.jsp` for parity.
- JS appends the “ Tel: …” segment and fixes “, <timestamp>” spacing; the new parameter is optional for rolling deploys with cached assets.
- Tests cover both numbers, each alone, all blank, null pharmacy, verbatim formatting, newline collapsing, and trimming when both numbers are padded.

<sup>Written for commit aeeb83844cc7993f202a0aa88ad78996ba978047. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/openo-beta/Open-O/pull/2494?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by Sourcery

Include composed pharmacy phone numbers in faxed prescription chart notes and printed Rx address blocks, while correcting spacing before the chart note timestamp.

New Features:
- Add pharmacy phone number to the "Rx faxed to" chart note when a faxed prescription is posted, using phone1/phone2 when available.

Bug Fixes:
- Prevent printed prescriptions from showing "Tel: null" or stray separators when pharmacy phone fields are missing.

Enhancements:
- Introduce a shared helper to compose pharmacy phone1/phone2 into a single display value reused across Rx views and notes.

Tests:
- Add unit tests covering pharmacy phone composition for null pharmacies, empty/whitespace values, newline collapsing, and mixed-format numbers.

### CODE REVIEW

File: src/test-modern/java/ca/openosp/openo/test/prescript/RxPharmacyDataPhoneTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    @Test
+    @DisplayName("should join both numbers with a single space when both are present")
+    void shouldJoinBothNumbers_whenBothPresent() {
+        String phone = RxPharmacyData.composePharmacyPhone(
+                pharmacyWithPhones("(416) 269-4820", "416-555-0000"));
+
+        assertThat(phone).isEqualTo("(416) 269-4820 416-555-0000");
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/test-modern/java/ca/openosp/openo/test/prescript/RxPharmacyDataPhoneTest.java:61]: Will be added in a commit soon

File: src/main/java/ca/openosp/openo/prescript/data/RxPharmacyData.java [hunk 2]
```diff
@@ -236,4 +237,34 @@ public void unlinkPharmacy(String pharmacyId, String demographicNo) {
...
+     * phone1 and phone2 with a single space, skipping whichever is absent so there is never a stray
+     * separator or a literal "null".
+     *
+     * <p>Numbers are returned verbatim apart from trimming and newline collapsing. Stored pharmacy
+     * phone numbers are unvalidated free text in mixed formats, so callers must not normalise
+     * them.</p>
+     *
+     * @param pharmacy PharmacyInfo the pharmacy, may be null
+     * @return String the joined phone numbers, or "" if none are on file
+     */
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/prescript/data/RxPharmacyData.java:254]: Will be added in a commit soon

File: src/main/java/ca/openosp/openo/rx/RxPrintPreview2Action.java [hunk 3]
```diff
@@ -557,11 +557,10 @@ private void setupPharmacyAddress(HttpServletRequest request, PharmacyInfo pharm
         addressJoiner.add(pharmacy.getName());
         addressJoiner.add(pharmacy.getAddress());
         addressJoiner.add(pharmacy.getCity() + ", " + pharmacy.getProvince() + " " + pharmacy.getPostalCode());
-        String tel = "Tel: " + pharmacy.getPhone1();
-        if (pharmacy.getPhone2() != null && !pharmacy.getPhone2().isEmpty()) {
-            tel += " " + pharmacy.getPhone2();
+        String tel = RxPharmacyData.composePharmacyPhone(pharmacy);
+        if (!tel.isEmpty()) {
+            addressJoiner.add("Tel: " + tel);
         }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/rx/RxPrintPreview2Action.java:563]: Fair, but I think this isn't needed at the moment, the JSP itself should already have the encoding, and it shouldn't be added into this change but in another PR if wanted

File: src/main/webapp/oscarRx/printRx/PrintPreview.js [hunk 4]
```diff
@@ -414,9 +414,13 @@ function printPaste2Parent(ctx, print, fax, pasteRx, rxPasteAsterisk, prefPharma
...
             text = "[Rx faxed to " + pharmacyName.replace(/\\n/g, '\n') + " Fax#: " + pharmacyFax;
 
+            if (pharmacyPhone && pharmacyPhone.trim() !== "") {
+                text += " Tel: " + pharmacyPhone.trim();
+            }
+
             text += " prescribed by " + prescribedBy;
 
-            text += "," + timeStamp + "]\n";
+            text += ", " + timeStamp + "]\n";
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/printRx/PrintPreview.js:423]: Good point on the contradiction, fixing the description rather than the code.

The separator change is deliberate and already disclosed under "Additional notes for reviewers": it restores a space that was dropped when this path was ported from ViewScript2.jsp, which still has it. It's isolated in its own commit (ee501a5760) specifically so it can be reverted independently of the feature if wanted.

What was wrong is the behaviour table and the "byte-identical" line, which were meant to say that no Tel: segment is added when there's no phone on file, not that the whole note is unchanged. Both reworded.

File: src/main/java/ca/openosp/openo/prescript/data/RxPharmacyData.java [hunk 5]
```diff
@@ -236,4 +237,35 @@ public void unlinkPharmacy(String pharmacyId, String demographicNo) {
...
+            return "";
+        }
+        StringJoiner phones = new StringJoiner(" ");
+        for (String phone : new String[]{pharmacy.getPhone1(), pharmacy.getPhone2()}) {
+            if (phone != null) {
+                // Collapse embedded newlines so the chart note stays on one line once encoded.
+                String cleaned = phone.replaceAll("[\\r\\n]+", " ").trim();
+                if (!cleaned.isEmpty()) {
+                    phones.add(cleaned);
+                }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/prescript/data/RxPharmacyData.java:266]: Declining the code change; the description was the imprecise part and I've fixed it.

The helper's javadoc already carves this out — "returned verbatim apart from trimming and newline collapsing" — and three of the eight tests assert both behaviours (shouldTrimBothNumbers_whenBothArePadded, shouldCollapseEmbeddedNewlines_soNoteStaysOnOneLine, and the whitespace-only case in shouldReturnEmpty). The PR body said "verbatim" without the carve-out, which is where the contradiction came from; now reworded.

Both normalisations are load-bearing. Without trim(), a whitespace-only phone field passes the empty guard and renders Tel: followed by blanks — the same dangling-label artifact as the Tel: null case this PR removes. Without the newline collapse, a CR/LF gets encoded to a \n escape that the JS parser materialises as a real line break, splitting the bracketed note across two lines.

Worth noting nothing is written back to the database — this only affects the composed display value.

**Review by D3V41 — APPROVED** [review]
Looks good
